cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetInstanceTable.java

```diff
@@ -0,0 +1,142 @@
+import co.cask.cdap.spi.data.table.field.Field;
+import co.cask.cdap.spi.data.table.field.Fields;
+import co.cask.cdap.spi.data.table.field.Range;
+import co.cask.cdap.store.StoreDefinition;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.gson.Gson;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Predicate;
+import javax.annotation.Nullable;
+
+/**
+ * Dataset instances metadata store
+ */
+public final class DatasetInstanceTable {
+  private static final Gson GSON = new Gson();
+
+  private StructuredTable table;
+
+  public DatasetInstanceTable(StructuredTableContext context) throws TableNotFoundException {
+    table = context.getTable(StoreDefinition.DatasetInstanceStore.DATASET_INSTANCES);
+  }
+
+  @Nullable
+  public DatasetSpecification get(DatasetId id) throws IOException {
+    Optional<StructuredRow> row =
+      table.read(
+        ImmutableList.of(Fields.stringField(StoreDefinition.DatasetInstanceStore.NAMESPACE_FIELD,
+                                            id.getParent().getEntityName()),
+                         Fields.stringField(StoreDefinition.DatasetInstanceStore.DATASET_FIELD, id.getEntityName())));
+    if (!row.isPresent()) {
+      return null;
+    }
+    return GSON.fromJson(
+      row.get().getString(StoreDefinition.DatasetInstanceStore.DATASET_METADATA_FIELD), DatasetSpecification.class);
+  }
+
+  public void write(NamespaceId namespaceId, DatasetSpecification instanceSpec) throws IOException {
+    Field<String> namespaceField = Fields.stringField(StoreDefinition.DatasetInstanceStore.NAMESPACE_FIELD,
+                                                      namespaceId.getEntityName());
+    Field<String> datasetField =
+      Fields.stringField(StoreDefinition.DatasetInstanceStore.DATASET_FIELD, instanceSpec.getName());
+    Field<String> metadataField =
+      Fields.stringField(StoreDefinition.DatasetInstanceStore.DATASET_METADATA_FIELD, GSON.toJson(instanceSpec));
+    table.upsert(ImmutableList.of(namespaceField, datasetField, metadataField));
+  }
+
+  public boolean delete(DatasetId datasetInstanceId) throws IOException {
+    if (get(datasetInstanceId) == null) {
+      return false;
+    }
+    table.delete(
+      ImmutableList.of(Fields.stringField(StoreDefinition.DatasetInstanceStore.NAMESPACE_FIELD,
+                                          datasetInstanceId.getParent().getEntityName()),
+                       Fields.stringField(StoreDefinition.DatasetInstanceStore.DATASET_FIELD,
+                                          datasetInstanceId.getEntityName())));
+    return true;
+  }
+
+  public Collection<DatasetSpecification> getAll(NamespaceId namespaceId) throws IOException {
```

Comment:
it would be good to have Javadoc for this method. The name of the method does not reveal that it filters out workflow-local datasets.